lib/puppet/pops/types/type_calculator.rb

```diff
@@ -582,115 +365,86 @@ def common_type(t1, t2)
 
     # when both are arrays, return an array with common element type
     if t1.is_a?(Types::PArrayType) && t2.is_a?(Types::PArrayType)
-      type = Types::PArrayType.new()
-      type.element_type = common_type(t1.element_type, t2.element_type)
-      return type
+      return Types::PArrayType.new(common_type(t1.element_type, t2.element_type))
     end
 
     # when both are hashes, return a hash with common key- and element type
     if t1.is_a?(Types::PHashType) && t2.is_a?(Types::PHashType)
-      type = Types::PHashType.new()
-      type.key_type = common_type(t1.key_type, t2.key_type)
-      type.element_type = common_type(t1.element_type, t2.element_type)
-      return type
+      key_type = common_type(t1.key_type, t2.key_type)
+      element_type = common_type(t1.element_type, t2.element_type)
+      return Types::PHashType.new(key_type, element_type)
     end
 
     # when both are host-classes, reduce to PHostClass[] (since one was not assignable to the other)
     if t1.is_a?(Types::PHostClassType) && t2.is_a?(Types::PHostClassType)
-      return Types::PHostClassType.new()
+      return Types::PHostClassType::DEFAULT
     end
 
     # when both are resources, reduce to Resource[T] or Resource[] (since one was not assignable to the other)
     if t1.is_a?(Types::PResourceType) && t2.is_a?(Types::PResourceType)
-      result = Types::PResourceType.new()
       # only Resource[] unless the type name is the same
-      if t1.type_name == t2.type_name then result.type_name = t1.type_name end
+      return t1.type_name == t2.type_name ?  Types::PResourceType.new(t1.type_name, nil) : Types::PResourceType::DEFAULT
       # the cross assignability test above has already determined that they do not have the same type and title
-      return result
     end
 
     # Integers have range, expand the range to the common range
     if t1.is_a?(Types::PIntegerType) && t2.is_a?(Types::PIntegerType)
-      t1range = from_to_ordered(t1.from, t1.to)
-      t2range = from_to_ordered(t2.from, t2.to)
-      t = Types::PIntegerType.new()
-      from = [t1range[0], t2range[0]].min
-      to = [t1range[1], t2range[1]].max
-      t.from = from unless from == Float::INFINITY
-      t.to = to unless to == Float::INFINITY
-      return t
+      return Types::PIntegerType.new([t1.from, t2.from].min, [t1.to, t2.to].max)
     end
 
     # Floats have range, expand the range to the common range
     if t1.is_a?(Types::PFloatType) && t2.is_a?(Types::PFloatType)
-      t1range = from_to_ordered(t1.from, t1.to)
-      t2range = from_to_ordered(t2.from, t2.to)
-      t = Types::PFloatType.new()
-      from = [t1range[0], t2range[0]].min
-      to = [t1range[1], t2range[1]].max
-      t.from = from unless from == Float::INFINITY
-      t.to = to unless to == Float::INFINITY
-      return t
+      return Types::PFloatType.new([t1.from, t2.from].min, [t1.to, t2.to].max)
```

Comment:
Both this and the corresponding Integer have a flaw (and so did the original). They both expand the range to include every value between min and max. That is not great. It would be better to have a Variant of integer ranges, and to replace a range with the merged if the found range is adjacent, otherwise it is added as a another range in the variant. Finally removing the variant if only containing one integer range. Thus, the result of [1,3,5, 2, 4] would be Integer[1,4], after having been Variant[Integer[1]], ... Variant[Integer[1], Integer[3], Integer[5]]... until '2' is merged, Variant[Integer[1,3], Integer[5]], then '4'.

Not worth doing this in this PR, but worth a ticket.
(OTOH, I could be reading this diff wrong - and what I am talking about is irrelevant because we do want generalization to reduce it to a wider (and "incorrect" range)).